// install-local/src/Options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Options {
    constructor(argv) {
        const args = argv // strip the "node install-local" part.
            .filter((_, i) => i > 1);
        this.dependencies = args
            .filter(arg => arg.substr(0, 1) !== '-');
        this.options = args
            .filter(arg => arg.substr(0, 1) === '-');
    }
    validate() {
        if (this.dependencies.length > 0 && this.targetSiblings) {
            return Promise.reject(`Invalid use of option --target-siblings. Cannot be used together with a dependency list`);
        }
        else if (this.targetSiblings && this.save) {
            return Promise.reject(`Invalid use of option --target-siblings. Cannot be used together with --save`);
        }
        else {
            return Promise.resolve();
        }
    }
    get help() {
        return this.flag('-h', '--help');
    }
    get targetSiblings() {
        return this.flag('-T', '--target-siblings');
    }
    get save() {
        return this.flag('-S', '--save');
    }
    flag(...options) {
        return options.some(_ => this.options.indexOf(_) >= 0);
    }
}
exports.Options = Options;
