// install-local/src/save.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("mz/fs"));
const path_1 = __importDefault(require("path"));
function saveIfNeeded(options) {
    return targets => {
        if (options.save) {
            return Promise.all(targets.map(save)).then(() => undefined);
        }
        else {
            return Promise.resolve();
        }
    };
}
exports.saveIfNeeded = saveIfNeeded;
function save(target) {
    const dependencies = target.packageJson.localDependencies || (target.packageJson.localDependencies = {});
    const dependenciesBefore = Object.assign({}, dependencies);
    target.sources.sort((a, b) => a.directory.localeCompare(b.directory))
        .forEach(source => dependencies[source.packageJson.name] = path_1.default.relative(target.directory, source.directory).replace(/\\/g, '/'));
    if (equals(dependencies, dependenciesBefore)) {
        return Promise.resolve();
    }
    else {
        return savePackageJson(target);
    }
}
function savePackageJson(target) {
    return fs_1.default.writeFile(path_1.default.resolve(target.directory, 'package.json'), JSON.stringify(target.packageJson, undefined, 2), { encoding: 'utf8' });
}
function equals(a, b) {
    const aNames = sortedNames(a);
    const bNames = sortedNames(b);
    if (aNames.length === bNames.length) {
        while (aNames.length) {
            if (!equalDependency(aNames.pop(), bNames.pop(), a, b)) {
                return false;
            }
        }
        return true;
    }
    return false;
}
function equalDependency(aKey, bKey, aDeps, bDeps) {
    return aKey === bKey && aKey && bKey && aDeps[aKey] === bDeps[bKey];
}
function sortedNames(subject) {
    return Object.keys(subject).sort();
}
